services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -267,18 +304,45 @@ private boolean isPushInTerminalState(Version targetVersion) {
       return false;
     }
 
-    // The store is eligible for a version swap if its push job is in terminal status. For a target region
-    // push, the parent version status is set to PUSHED in getOfflinePushStatus when this happens or KILLED if the push
-    // failed. PUSHED represents when a push successfully completes in all regions and KILLED represents when a push
-    // fails in
-    // 1+ regions. KILLED is still eligible for a version swap because some non target regions may have succeeded, and
-    // we
-    // need to perform a version swap for those regions
-    if (targetVersion.getStatus() != VersionStatus.PUSHED && targetVersion.getStatus() != VersionStatus.KILLED) {
-      return false;
+    // The version could've been manually rolled forward in non target regions. If that is the case, we should check if
+    // we emitted any stalled
+    // metric for it and remove it if so
+    if (stalledVersionSwapSet.contains(storeName)) {
+      boolean didPushCompleteInNonTargetRegions =
+          isVersionOnlineInRegions(nonTargetRegions, clusterName, storeName, targetVersion.getNumber());
+      if (didPushCompleteInNonTargetRegions) {
+        stalledVersionSwapSet.remove(storeName);
+        deferredVersionSwapStats.recordDeferredVersionSwapStalledVersionSwapSensor(stalledVersionSwapSet.size());
+      }
     }
 
-    return true;
+    switch (targetVersion.getStatus()) {
+      case STARTED:
+        // Because the parent status is updated when we poll for the job status, the parent status will not always be an
+        // accurate representation of push status if vpj runs away
+        Set<String> targetRegions = RegionUtils.parseRegionsFilterList(targetRegionsString);
+        boolean didPushCompleteInTargetRegions =
+            isVersionOnlineInRegions(targetRegions, clusterName, storeName, targetVersion.getNumber());
+        if (didPushCompleteInTargetRegions) {
+          deferredVersionSwapStats.recordDeferredVersionSwapRunawayVpjSensor();
+          String message =
+              "Push completed in target regions, but vpj ran away. Continuing with deferred swap for store: "
```

Comment:
nit: lets just say "...but parent version is still in STARTED status..." as we not sure exactly what is happening here